transport/src/main/java/io/netty/channel/AbstractCoalescingBufferQueue.java

```diff
@@ -0,0 +1,267 @@
+        if (bufAndListenerPairs.isEmpty()) {
+            return removeEmptyValue();
+        }
+        bytes = Math.min(bytes, readableBytes);
+
+        ByteBuf toReturn = null;
+        int originalBytes = bytes;
+        for (;;) {
+            Object entry = bufAndListenerPairs.poll();
+            if (entry == null) {
+                break;
+            }
+            if (entry instanceof ChannelFutureListener) {
+                aggregatePromise.addListener((ChannelFutureListener) entry);
+                continue;
+            }
+            ByteBuf entryBuffer = (ByteBuf) entry;
+            if (entryBuffer.readableBytes() > bytes) {
+                // Add the buffer back to the queue as we can't consume all of it.
+                bufAndListenerPairs.addFirst(entryBuffer);
+                if (bytes > 0) {
+                    // Take a slice of what we can consume and retain it.
+                    ByteBuf next = entryBuffer.readRetainedSlice(bytes);
+                    toReturn = toReturn == null ? composeFirst(alloc, next) : compose(alloc, toReturn, next);
+                    bytes = 0;
+                }
+                break;
+            } else {
+                bytes -= entryBuffer.readableBytes();
+                toReturn = toReturn == null ? composeFirst(alloc, entryBuffer) : compose(alloc, toReturn, entryBuffer);
+            }
+        }
+        readableBytes -= originalBytes - bytes;
+        assert readableBytes >= 0;
+        return toReturn;
+    }
+
+    /**
+     * The number of readable bytes.
+     */
+    public int readableBytes() {
+        return readableBytes;
+    }
+
+    /**
+     * Are there pending buffers in the queue.
+     */
+    public boolean isEmpty() {
+        return bufAndListenerPairs.isEmpty();
+    }
+
+    /**
+     *  Release all buffers in the queue and complete all listeners and promises.
+     */
+    public void releaseAndFailAll(ChannelOutboundInvoker invoker, Throwable cause) {
+        releaseAndCompleteAll(invoker.newFailedFuture(cause));
+    }
+
+    /**
+     * Copy all pending entries in this queue into the destination queue.
+     * @param dest to copy pending buffers to.
+     */
+    public void copyTo(AbstractCoalescingBufferQueue dest) {
+        dest.bufAndListenerPairs.addAll(bufAndListenerPairs);
+        dest.readableBytes += readableBytes;
+    }
+
+    /**
+     * Writes all remaining elements in this queue.
+     * @param ctx The context to write all elements to.
+     */
+    public void writeAndRemoveAll(ChannelHandlerContext ctx) {
+        readableBytes = 0;
+        Throwable pending = null;
+        ByteBuf previousBuf = null;
+        for (;;) {
+            Object entry = bufAndListenerPairs.poll();
+            if (entry == null) {
+                if (previousBuf != null) {
+                    ctx.write(previousBuf, ctx.voidPromise());
+                }
+                break;
+            }
+            try {
```

Comment:
Why the try block starts not from start of loop? We can miss `pending` if the thrown will be above.